### PR TITLE
app/helpers i18n

### PR DESCRIPTION
**Pull request by yheuhtozr**
## What type of PR is this? (check all applicable)

- [X] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

I18n of ruby code under app/helpers. Involving some deconstantization as needed.

## Related Tickets & Documents

#14888

## QA Instructions, Screenshots, Recordings

_Please replace this line with instructions on how to test your changes, a note
on the devices and browsers this has been tested on, as well as any relevant
images for UI changes._

### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://developers.forem.com/frontend/accessibility)._

## Added/updated tests?

- [ ] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [X] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

**Comment by yheuhtozr** [comment]
A Travis job failed for some reason while all tests inside passed.

File: app/helpers/application_helper.rb [hunk 1]
```diff
@@ -1,27 +1,31 @@
 module ApplicationHelper
-  # rubocop:disable Style/OpenStructUse, Performance/OpenStruct
   USER_COLORS = ["#19063A", "#dce9f3"].freeze
 
-  DELETED_USER = OpenStruct.new(
```
Comment (citizen428) [inline app/helpers/application_helper.rb:5]: Couldn't we have kept this constant around and just changed `name`? Given that we can't get new translations without deploying anyway, I don't think there's anything wrong with this approach unless I'm missing something?

File: app/helpers/application_helper.rb [hunk 2]
```diff
@@ -1,27 +1,31 @@
...
-    name: "[Deleted User]",
-    summary: nil,
-    twitter_username: nil,
-    github_username: nil,
-  )
-  # rubocop:enable Style/OpenStructUse, Performance/OpenStruct
-
   LARGE_USERBASE_THRESHOLD = 1000
 
-  SUBTITLES = {
```
Comment (citizen428) [inline app/helpers/application_helper.rb:18]: Same as above, we can still keep a constant hash around, but the values will now come via I18n.

File: app/helpers/articles_helper.rb [hunk 3]
```diff
@@ -1,14 +1,12 @@
 module ArticlesHelper
-  DASHBOARD_POSTS_SORT_OPTIONS = [
```
Comment (citizen428) [inline app/helpers/articles_helper.rb:2]: Same comment as the two above.

File: app/helpers/articles_helper.rb [hunk 4]
```diff
@@ -47,8 +52,8 @@ def should_show_crossposted_on?(article)
   end
 
   def get_host_without_www(url)
-    url = "http://#{url}" if URI.parse(url).scheme.nil?
-    host = URI.parse(url).host.downcase
+    url = "http://#{url}" if Addressable::URI.parse(url).scheme.nil?
```
Comment (citizen428) [inline app/helpers/articles_helper.rb:55]: I have no problem with it, but why this change that is not I18n related?

File: app/helpers/authentication_helper.rb [hunk 5]
```diff
@@ -27,10 +27,10 @@ def signed_up_with(user = current_user)
     providers = Authentication::Providers.enabled_for_user(user)
 
     # If the user did not authenticate with any provider, they signed up with an email.
-    auth_method = providers.any? ? providers.map(&:official_name).to_sentence : "Email & Password"
-    verb = providers.size > 1 ? "any of those" : "that"
+    auth_method = providers.any? ? providers.map(&:official_name).to_sentence : I18n.t("helpers.authentication_helper.email_password") # rubocop:disable Layout/LineLength
```
Comment (citizen428) [inline app/helpers/authentication_helper.rb:30]: I don't think the `rubocop:disable` should go to the end of the line here on 2 consecutive lines, please use this form instead:

```ruby
# rubocop:disable Layout/LineLength
auth_method = ...
verb = ...
# rubocop:enable Layout/LineLength
```

File: app/helpers/authentication_helper.rb [hunk 6]
```diff
@@ -27,10 +27,10 @@ def signed_up_with(user = current_user)
...
+    verb = providers.size > 1 ? I18n.t("helpers.authentication_helper.any_of_those") : I18n.t("helpers.authentication_helper.that") # rubocop:disable Layout/LineLength
 
-    "Reminder: you used #{auth_method} to authenticate your account, so please use #{verb} to sign in if prompted."
+    I18n.t("helpers.authentication_helper.reminder", method: auth_method, dem: verb)
```
Comment (citizen428) [inline app/helpers/authentication_helper.rb:33]: `dem`?

File: app/helpers/comments_helper.rb [hunk 7]
```diff
@@ -23,16 +23,19 @@ def commentable_author_is_op?(commentable, comment)
...
   end
 
   def tree_for(comment, sub_comments, commentable)
     nested_comments(tree: { comment => sub_comments }, commentable: commentable, is_view_root: true)
   end
 
   def should_be_hidden?(comment, root_comment)
-    # when opened by a permalink + root comment is hidden => show root comment and its descendants
-    comment.hidden_by_commentable_user && comment != root_comment && !root_comment&.hidden_by_commentable_user
+    comment.hidden_by_commentable_user && comment != root_comment
```
Comment (citizen428) [inline app/helpers/comments_helper.rb:38]: This lost one condition, was this on purpose?

File: app/helpers/application_helper.rb (hunk 1 above)
Comment (yheuhtozr) [reply app/helpers/application_helper.rb:5]: Maybe this can work?
``` ruby
  DELETED_USER = OpenStruct.new(
    id: nil,
    darker_color: Color::CompareHex.new(USER_COLORS).brightness,
    username: "[deleted user]",
    name: -> { I18n.t("helpers.application_helper.deleted_user") },
    summary: nil,
    twitter_username: nil,
    github_username: nil,
  )
```

File: app/helpers/application_helper.rb (hunk 2 above)
Comment (yheuhtozr) [reply app/helpers/application_helper.rb:18]: Like this? (I didn't really test it out though)
``` ruby
  SUBTITLES = {
    "week" => -> { I18n.t("helpers.application_helper.subtitle.week") },
    "month" => -> { I18n.t("helpers.application_helper.subtitle.month") },
    "year" => -> { I18n.t("helpers.application_helper.subtitle.year") },
    "infinity" => -> { I18n.t("helpers.application_helper.subtitle.infinity") },
    "latest" => -> { I18n.t("helpers.application_helper.subtitle.latest") }
  }.freeze
```

File: app/helpers/articles_helper.rb (hunk 3 above)
Comment (yheuhtozr) [reply app/helpers/articles_helper.rb:2]: Enclosing i18n with lambda like above?

File: app/helpers/articles_helper.rb (hunk 4 above)
Comment (yheuhtozr) [reply app/helpers/articles_helper.rb:55]: In some sense it does. I believe they are here after I found they didn't handle non-ASCII path very well.

File: app/helpers/authentication_helper.rb (hunk 6 above)
Comment (yheuhtozr) [reply app/helpers/authentication_helper.rb:33]: https://en.wikipedia.org/wiki/Demonstrative
Any suggestions are welcome but I don't think it is `verb` 😂

File: app/helpers/comments_helper.rb (hunk 7 above)
Comment (yheuhtozr) [reply app/helpers/comments_helper.rb:38]: Oops, thank you!

File: app/helpers/application_helper.rb (hunk 1 above)
Comment (citizen428) [reply app/helpers/application_helper.rb:5]: Yes, that's exactly what I meant 💯

File: app/helpers/application_helper.rb (hunk 2 above)
Comment (citizen428) [reply app/helpers/application_helper.rb:18]: Ruby class bodies get evaluated on application start, so the values can be normal method calls:

```ruby
SUBTITLES = {
    "week" => I18n.t("helpers.application_helper.subtitle.week"),
    ...
```

File: app/helpers/articles_helper.rb (hunk 3 above)
Comment (citizen428) [reply app/helpers/articles_helper.rb:2]: No lambdas needed.

File: app/helpers/authentication_helper.rb (hunk 6 above)
Comment (citizen428) [reply app/helpers/authentication_helper.rb:33]: Gotcha, makes sense. In that case we should probably change the variable name too. I know it's a silly name but at least once we used something like `the_thing` in the codebase. 🤣

File: app/helpers/application_helper.rb (hunk 2 above)
Comment (yheuhtozr) [reply app/helpers/application_helper.rb:18]: Perhaps I did it wrong in the past, but I'm afraid, if it only evaluated once on the app's start, the value will never change until the next startup. That means once it started (i.e. on the default locale), it sticks with it no matter viewers' locales are.

Comment (yheuhtozr) [reply app/helpers/application_helper.rb:18]: As far as I understand, Rails doesn't fork a new app instance with each language, so the locale when this code is evaluated for the first time should be used forever.

Comment (citizen428) [reply app/helpers/application_helper.rb:18]: Yes, you're right, I was sure I was missing something here.

File: app/helpers/application_helper.rb (hunk 1 above)
Comment (yheuhtozr) [reply app/helpers/application_helper.rb:5]: I tried a bit in this direction and I think it must look like this to make other codes believe these constants exist while being i18n-friendly. I have no idea if I have to brace for someone trying to override the `:user` property for this case, though.

``` ruby
  DELETED_USER = Class.new(OpenStruct) do
    def initialize
      super
      self[:id] = nil
      self[:darker_color] = Color::CompareHex.new(USER_COLORS).brightness
      self[:username] = "[deleted user]"
      self[:summary] = nil
      self[:twitter_username] = nil
      self[:github_username] = nil
    end

    def [](key)
      return name if key.to_s == "name"
      super
    end
  
    def name
      I18n.t("helpers.application_helper.deleted_user")
    end
  end.new
```

Which do you think better, to wrap and simulate any of those referred constants or to methodify them and change all their callers? I may be overlooking some easier ways, any input is appreciated.

Comment (citizen428) [reply app/helpers/application_helper.rb:5]: I think it's ok to go for a method, I thought about it the wrong way yesterday as we discussed.

Comment (citizen428) [reply app/helpers/application_helper.rb:5]: This part is currently reworked in https://github.com/forem/forem/pull/16070. I think we should wait for that to get merged, then pull in the changes and override the `name` method to return the translation instead.

**Review by citizen428 — APPROVED** [review]

**Review by jeremyf — APPROVED** [review]
